presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java

```diff
@@ -108,10 +110,9 @@ public Result apply(LateralJoinNode lateralJoinNode, Captures captures, Context
                 .recurseOnlyWhen(ProjectNode.class::isInstance)
                 .removeFirst();
 
-        Range<Long> subqueryCardinality = extractCardinality(rewrittenSubquery, context.getLookup());
-        boolean producesAtMostOneRow = Range.closed(0L, 1L).encloses(subqueryCardinality);
+        boolean producesAtMostOneRow = isAtMostScalar(rewrittenSubquery, context.getLookup());
         if (producesAtMostOneRow) {
-            boolean producesSingleRow = Range.singleton(1L).encloses(subqueryCardinality);
+            boolean producesSingleRow = isScalar(rewrittenSubquery, context.getLookup());
```

Comment:
I didn't notice that we're doing visiting twice here. Apologies! I would prefer that the code stays as it was, i.e. we `extractCardinality` once, not twice.

Reply:
Have reverted it back.